Use the QuarkusForkJoinWorkerThreadFactory for fork-join pools

Sanne opened the pull request:
An experiment to see if we can better control the Classloaders in the Fork/Join initiated threads.

@stuartwdouglas WDYT of the approach?  For dev-mode we'd likely want a slightly different implementation, so to be able to capture all threads when they are started, and then have the ability to swap the CL.

radcortez commented:
Gave it a try with https://github.com/quarkusio/quarkus/issues/20067 and it seems to fix the issue. I think it makes sensed to move to this approach.

radcortez commented on independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/forkjoin/QuarkusForkJoinWorkerThreadFactory.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinWorkerThread;
+
+public final class QuarkusForkJoinWorkerThreadFactory implements ForkJoinPool.ForkJoinWorkerThreadFactory {
```
Maybe we want to apply this to `QuarkusDevModeTest`, `QuarkusUnitTest` and `QuarkusProdModeTest`?

Sanne replied:
Probably? I'm not familiar with these

I do suspect thought that we'll want to have a slightly different implementation for "real prod mode" than for tests

radcortez commented:
@Sanne @stuartwdouglas are we ok to move forward with this?

Sanne commented:
Hi @radcortez - yes sorry I resumed yesterday with this.

I was trying to add some tests, but as I worked on it I realized my new integration tests were pointless, as it affects production mode only so @QuarkusTest isn't affected by the issue.

Do we have any existing pattern / utility to test such things? Any suggestion?

radcortez commented:
Have a look here into how the `maven` integration tests work: https://github.com/quarkusio/quarkus/blob/main/test-framework/maven/src/main/java/io/quarkus/maven/it/RunAndCheckMojoTestBase.java#L21.

Not exactly what we need, because it runs `quarkus:dev`, but maybe it can be adjusted (or copied) to run in prod mode. I'm not sure if we have anything else.

I guess that it should be also possible to run the prod jar in `pre-integrations-tests` and then run tests with `failsafe`.

Sanne commented:
thanks for the pointers!

Sanne committed:
Introduce a custom QuarkusForkJoinWorkerThread to controll the classloader in the Fork/Join common pool

Sanne committed:
Introduce integration tests to cover the production classloader

radcortez commented:
Wondering if the test fails are related, but at first glance it doesn't seem like it.

stuartwdouglas committed:
Move test to test-extension

This is where core functionality is usually tested, and removes the need
for another maven module.

radcortez commented:
> @radcortez looks like we're ready to merge; is there any open issue this should be linked to?

I'm not aware of any other options issues. Please go ahead. Thanks!
